### ISSUE TITLE
Make Sense code prettier and faster

### ISSUE THREAD

**Issue by Woseseltops** [Signbank/Global-signbank#1047]
@susanodd thinks she might be able to make the `Sense` model prettier and faster simultaneously, this is the issue to give that a try

**Comment by susanodd** [comment]
@Jetske if you want to join in, the above code is very strange that it results in 'Nederlands' and 'Engels' since the interface language is EN when it is called and EN is the default language of the model translations.

In the templates, we looked at this before because the display order ends up different depending on what the interface language is set to (since they're not sorted).

You've made dictionaries based on strings where the 'name' appears.
I'm seeing if it will improve efficiency if the query is evaluated in python rather than in the template.
But to my surprise it's obtaining the wrong language name (and not the default model translation).

**Comment by susanodd** [comment]
I rewrote the main functionality of the sense methods by adding parameters so the various similar methods can share code.

https://github.com/Signbank/Global-signbank/blob/b13adcd429ee8da987ee847b87485adff16d895c/signbank/dictionary/models.py#L722-L770

**Comment by vanlummelhuizen** [comment]
> It has used the model translation of `language__name` for Dutch, not English.
> Why is it doing this?

I found that the -untranslated- `name` field for Language objects 'English', 'Dutch' and 'French' contain the Dutch translation: Engels, Nederlands and Frans. 

Apparently, the Django-model-translation is not in play here, so the default `name` field is used, instead of `name_en` for instance. 

My assessment is that if you do `.values('language__name'` you access the database field without Django-modeltranslation doing any form of translation. So if you do `.values('language__name_en'` you would get the English translation of the Language name.

@susanodd I cannot find the code from your comment anywhere in the code base., so I am unable to test using a browser (I tested using the Django shell).

**Comment by vanlummelhuizen** [comment]
> My assessment is that if you do `.values('language__name'` you access the database field without Django-modeltranslation doing any form of translation. So if you do `.values('language__name_en'` you would get the English translation of the Language name.

See also https://github.com/deschler/django-modeltranslation/issues/409

**Comment by Jetske** [comment]
Do you agree that the senses are now unshared, thus: can we delete that code that is activated by the setting SHARE_SENSES, including that setting itself? 
It is still a many to many model so sharing can always be added in later on (e.g. through the modal that shows similar senses), but the code I originally wrote for shared senses won't work anymore after all the changes. Deleting that would make the code much more readable.

**Comment by susanodd** [comment]
> 

Yes, they are unshared. I agree.

I think the new feature with links is more user-friendly so the user can actually check whether the senses are to be shared or if the keywords need to be refined and disambiguated.

**Comment by susanodd** [comment]
> > My assessment is that if you do `.values('language__name'` you access the database field without Django-modeltranslation doing any form of translation. So if you do `.values('language__name_en'` you would get the English translation of the Language name.
> 
> See also [deschler/django-modeltranslation#409](https://github.com/deschler/django-modeltranslation/issues/409)

Okay, so probably not to do it that way, since otherwise the method would need a parameter for the 'en' or 'nl'.

**Comment by susanodd** [comment]
> > It has used the model translation of `language__name` for Dutch, not English.
> > Why is it doing this?
> 
> I found that the -untranslated- `name` field for Language objects 'English', 'Dutch' and 'French' contain the Dutch translation: Engels, Nederlands and Frans.
> 
> Apparently, the Django-model-translation is not in play here, so the default `name` field is used, instead of `name_en` for instance.
> 
> My assessment is that if you do `.values('language__name'` you access the database field without Django-modeltranslation doing any form of translation. So if you do `.values('language__name_en'` you would get the English translation of the Language name.
> 
> @susanodd I cannot find the code from your comment anywhere in the code base., so I am unable to test using a browser (I tested using the Django shell).

Ooh, I guess it got removed when I rewrote it, so you can't see the original.
I modified it to be 'language' instead of 'language_name' because it wasn't supposed to pick the wrong language.

Maybe if you look at the pull request, where the newer code is, that's kind of programming around the problem by using a lookup.
(I'm not entirely sure "when" Django figures out what language.name is.)
(Or is is something like lambda language. language.name)

**Comment by susanodd** [comment]
@vanlummelhuizen I created a lookup this way to avoid getting the wrong language:

https://github.com/Signbank/Global-signbank/blob/e0055ec1955a5b73d7a9f24fb62726e5539daf32/signbank/dictionary/models.py#L729-L737

Then this is what is returned:
https://github.com/Signbank/Global-signbank/blob/e0055ec1955a5b73d7a9f24fb62726e5539daf32/signbank/dictionary/models.py#L739-L754

It uses the "lookup".

Is it so that the "language.name" has not been evaluated yet?

**Comment by susanodd** [comment]
I am using values to force it to not make unevaluated queries. (So it evaluates them in Python rather than putting a query in the template.)

**Comment by susanodd** [comment]
I don't always know when to use prefetch_related or select_related.
But this did become a problem in the Relations View of a gloss. The template did not evaluate the things that were related as expected. (It had already converted some "objects" to their string representation and did not allow to apply a method to that object. I had to rewrite the code to move it to the python for the view.)
(That was probably an example of due to failure to select or prefetch whatever the related object was.)

**Comment by susanodd** [comment]
I merged the improvements to master and deployed them.

**Comment by susanodd** [comment]
> @susanodd thinks she might be able to make the `Sense` model prettier and faster simultaneously, this is the issue to give that a try

I rewrote the methods to evaluate to values in Python rather than return a query to the template. I replaced similar methods that had duplicate code (6 methods) with a single method (1 method) that has parameters for the join character and for whether to include empty languages in the dictionary.
The GlossDetailView uses two new filters to pass the parameters in the different places the methods are called.
